as/plugins/org.jboss.ide.eclipse.as.wtp.core/src/org/jboss/ide/eclipse/as/wtp/core/server/publish/BinaryModulePublishRunner.java

```diff
@@ -0,0 +1,218 @@
+		} else {
+			publishModuleIncremental(delta, ms);
+		}
+		return ms;
+	}
+	public void publishModuleIncremental(IModuleResourceDelta[] delta, MultiStatus ms) throws CoreException {
+		for( int i = 0; i < delta.length; i++ ) {
+			int kind = delta[i].getKind();
+			IModuleResource mr = delta[i].getModuleResource();
+			if( kind == IModuleResourceDelta.REMOVED) {
+				removeMembers(new IModuleResource[] {mr}, ms);
+			} else if( kind == IModuleResourceDelta.CHANGED || kind == IModuleResourceDelta.ADDED) {
+				if( mr instanceof IModuleFolder ) {
+					IModuleResourceDelta[] children = delta[i].getAffectedChildren();
+					publishModuleIncremental(children, ms);
+				} else {
+					copyOneResource(mr, ms);
+				}
+			}
+		}
+	}
+	
+	/*
+	 * Remove the module. Then do a full publish
+	 */
+	public MultiStatus publishModuleFull() throws CoreException {
+		MultiStatus status = (MultiStatus)removeDeletedBinaryModule();
+		if( isNestedSingleResourceBinaryModule()) {
+			copyOneResourceSingleBinary(all[0], status);
+		} else {
+			publishModuleMembers(all, status);
+		}
+		return status;
+	}
+	
+	
+	private boolean isNestedSingleResourceBinaryModule() {
+		if( module.length > 1 && all.length == 1 && all[0] instanceof IModuleFile) {
+			return true;
+		}
+		return false;
+	}
+	private void publishModuleMembers(IModuleResource[] members, MultiStatus ms) throws CoreException {
+		for( int i = 0; i < members.length; i++ ) {
+			IModuleResource mr = members[i];
+			if( mr instanceof IModuleFolder ) {
+				IModuleResource[] children = ((IModuleFolder)mr).members();
+				publishModuleMembers(children, ms);
+			} else {
+				copyOneResource(mr, ms);
+			}
+		}
+	}
+
+	private void copyOneResource(IModuleResource r, MultiStatus ms) throws CoreException {
+		copyOneResource(r, r.getName(), ms);
+	}
+	
+
+	private void copyOneResourceSingleBinary(IModuleResource r, MultiStatus ms) throws CoreException {
+		copyOneResource(r, archiveDestinationWithName.lastSegment(), ms);
+	}
+	
+	private void copyOneResource(IModuleResource r, String lastSegmentName, MultiStatus ms) throws CoreException {
+		File f = ModuleResourceUtil.getFile(r);
+		IPath folder = archiveDestination.append(r.getModuleRelativePath());
+		IPath dest = folder.append(lastSegmentName);
+		IStatus s1 = fc.makeDirectoryIfRequired(folder, new NullProgressMonitor());
+		IStatus result = null;
+		if( s1 != null && !s1.isOK()) {
+			result = s1;
+		} else {
+			IStatus s2 = fc.copyFile(f, dest, new NullProgressMonitor());
+			result = s2;
+		}
+		if( result != null )
+			ms.add(result);
+	}
+
+	
+	private IModuleResource[] getAllResources(IModuleResourceDelta[] delta, int kind, boolean depthFirst) {
+		if( delta == null )
+			return new IModuleResource[0];
+		
+		ArrayList<IModuleResource> list = new ArrayList<IModuleResource>();
+		for( int i = 0; i < delta.length; i++ ) {
+			IModuleResource r = delta[i].getModuleResource();
+			IModuleResourceDelta[] children = delta[i].getAffectedChildren();
```

Comment:
delta[i] is used several times, what do you think to extract it as local variable?

Reply:
Nah... array indexing is super fast

Reply:
my concern wasn't about performance but readability

Reply:
ok, I tried to provide an extracted local variable but the only name I found is modifiedModuleresource or moduleResourceDelta which is not increasing readability.
@robstryker do you have better name in mind describing this variable? If not, we will need to leave it as-is.